File: matrix/matrix.go
package matrix

import (
	"fmt"

	"gitlab.ocnr.org/apps/minesweeper/field"
)

type Matrix []Row

type Solution struct {
	Flagged, Revealed []*field.Space
}

// NewMatrix creates a new matrix of revealed edge spaces (rows) and unknown edge
// spaces (cols).
func NewMatrix(f *field.Field) Matrix {
	var matrix Matrix
	revealedEdgeSpaces := f.RevealedEdgeSpaces()
	unknownEdgeSpaces := f.UnknownEdgeSpaces()
	for _, revealedEdgeSpace := range revealedEdgeSpaces {
		var row Row
		for _, unknownEdgeSpace := range unknownEdgeSpaces {
			value := 0.0
			if revealedEdgeSpace.UnknownNeighbors()[unknownEdgeSpace] {
				value = 1
			}
			row = append(row, &Cell{value, unknownEdgeSpace, field.Unknown})
		}
		rhs := float64(revealedEdgeSpace.MineNeighborCount() - revealedEdgeSpace.FlaggedNeighborCount())
		row = append(row, &Cell{rhs, revealedEdgeSpace, field.Unknown})
		matrix = append(matrix, row)
	}
	matrix.reduce()
	return matrix.removeUnconstrained()
}

// SplitCoupled creates new matrices from rows that do not share constraints.
func (m Matrix) SplitCoupled() []Matrix {
	var matrices []Matrix
	src := m.copy()
	for len(src) > 0 {
		dst := Matrix{src[0]}
		src = src.removeRow(src[0].Rhs().space)
		src, dst = coupledRows(src, dst)
		dst = dst.removeUnconstrained()
		matrices = append(matrices, dst)
	}
	return matrices
}

// Resolve resolves states determined by constraints, inspired by
// https://massaioli.wordpress.com/2013/01/12/solving-minesweeper-with-matricies/.
func (m Matrix) Resolve() (Matrix, error) {
	matrix := m.copy()
	for _, row := range matrix {
		lhs := row.ConstrainedLhs()
		lhsNegativeSum := 0.0
		lhsPositiveSum := 0.0
		for _, cell := range lhs {
			if cell.value < 0 {
				lhsNegativeSum += cell.value
				continue
			}
			lhsPositiveSum += cell.value
		}
		rhs := row.Rhs().value
		if rhs == lhsNegativeSum {
			for _, cell := range lhs {
				if cell.value < 0 {
					if err := matrix.flag(cell.space); err != nil {
						return matrix, err
					}
				} else {
					if err := matrix.reveal(cell.space); err != nil {
						return matrix, err
					}
				}
			}
			continue
		}
		if rhs == lhsPositiveSum {
			for _, cell := range lhs {
				if cell.value < 0 {
					if err := matrix.reveal(cell.space); err != nil {
						return matrix, err
					}
				} else {
					if err := matrix.flag(cell.space); err != nil {
						return matrix, err
					}
				}
			}
			continue
		}
	}
	return matrix, nil
}

// Solve returns a list of all possible solutions for the matrix.
func (m Matrix) Solve(visualize bool) []*Solution {
	space := m.mostConstrainedSpace()
	if space == nil {
		solution := &Solution{[]*field.Space{}, []*field.Space{}}
		for _, cell := range m[0].Lhs() {
			switch cell.state {
			case field.Flagged:
				solution.Flagged = append(solution.Flagged, cell.space)
			case field.Revealed:
				solution.Revealed = append(solution.Revealed, cell.space)
			case field.Unknown:
				// Ignore
			}
		}
		if visualize {
			m.Print()
		}
		return []*Solution{solution}
	}
	var solutions []*Solution
	reveal := m.copy()
	if err := reveal.reveal(space); err == nil {
		if reveal, err = reveal.Resolve(); err == nil {
			solutions = append(solutions, reveal.Solve(visualize)...)
		}
	}
	flag := m.copy()
	if err := flag.flag(space); err == nil {
		if flag, err = flag.Resolve(); err == nil {
			solutions = append(solutions, flag.Solve(visualize)...)
		}
	}
	return solutions
}

// ConstrainedSpaces returns a map of the constrained spaces.
func (m Matrix) ConstrainedSpaces() map[*field.Space]bool {
	spaces := make(map[*field.Space]bool)
	for _, row := range m {
		for _, cell := range row.ConstrainedLhs() {
			spaces[cell.space] = true
		}
	}
	return spaces
}

// Print prints the matrix to stdout.
func (m Matrix) Print() {
	if len(m) == 0 {
		fmt.Println("(empty matrix)")
	}
	for rowIndex, row := range m {
		if rowIndex == 0 {
			fmt.Print(" ")
			for _, cell := range row.Lhs() {
				fmt.Printf("%4d ", cell.space.Index())
			}
			fmt.Println()
			for _, cell := range row.Lhs() {
				switch cell.state {
				case field.Unknown:
					fmt.Print("    U")
				case field.Flagged:
					fmt.Print("    F")
				case field.Revealed:
					fmt.Print("    R")
				}
			}
			fmt.Println()
		}
		for cellIndex, cell := range row {
			if cellIndex%len(row) == len(row)-1 {
				fmt.Printf(" | %4s %d", fmt.Sprintf("%1.1f", cell.value), cell.space.Index())
			} else {
				fmt.Printf(" %4s", fmt.Sprintf("%1.1f", cell.value))
			}
		}
		fmt.Println()
	}
	fmt.Println()
}

// mostConstrainedSpace returns the space with the most constraints.
func (m Matrix) mostConstrainedSpace() *field.Space {
	var mostConstrainedSpace *field.Space
	mostConstrainedCount := 0
	constraintCounts := make(map[*field.Space]int)
	for _, row := range m {
		for _, cell := range row.ConstrainedLhs() {
			constraintCounts[cell.space]++
			if constraintCounts[cell.space] > mostConstrainedCount {
				mostConstrainedSpace = cell.space
				mostConstrainedCount = constraintCounts[cell.space]
			}
		}
	}
	return mostConstrainedSpace
}

// unconstrainedSpaces returns a map of the unconstrained spaces.
func (m Matrix) unconstrainedSpaces() map[*field.Space]bool {
	spaces := make(map[*field.Space]bool)
	for _, cell := range m[0].Lhs() {
		spaces[cell.space] = true
	}
	for _, row := range m {
		for _, cell := range row.ConstrainedLhs() {
			delete(spaces, cell.space)
		}
	}
	return spaces
}

// zeroRows returns all rows with all-zero left-hand-sides.
func (m Matrix) zeroRows() map[*field.Space]bool {
	zeroRows := make(map[*field.Space]bool)
	for _, row := range m {
		if len(row.ConstrainedLhs()) == 0 {
			zeroRows[row.Rhs().space] = true
		}
	}
	return zeroRows
}

// coupledRows returns a matrix consisting of rows from source which are coupled
// to destination.
func coupledRows(src Matrix, dst Matrix) (Matrix, Matrix) {
	if len(src) == 0 {
		return src, dst
	}
	dstConstrainedSpaces := dst.ConstrainedSpaces()
	for _, srcRow := range src {
		for _, srcCell := range srcRow.ConstrainedLhs() {
			if dstConstrainedSpaces[srcCell.space] {
				dst = append(dst, srcRow)
				src = src.removeRow(srcRow.Rhs().space)
				return coupledRows(src, dst)
			}
		}
	}
	return src, dst
}

// reduce transforms the matrix to RREF. This algorithm is pilfered from
// https://rosettacode.org/wiki/Reduced_row_echelon_form.
func (m Matrix) reduce() {
	if len(m) == 0 {
		return
	}
	lead := 0
	rowCount := len(m)
	colCount := len(m[0])
	for r := range m {
		if lead >= colCount {
			return
		}
		i := r
		for m[i][lead].value == 0 {
			i++
			if i == rowCount {
				i = r
				lead++
				if lead == colCount {
					return
				}
			}
		}
		m[i], m[r] = m[r], m[i]
		div := m[r][lead].value
		for j := range m[r] {
			m[r][j].value /= div
		}
		for j := range m {
			if j != r {
				sub := m[j][lead].value
				for k := range m[r] {
					m[j][k].value -= sub * m[r][k].value
				}
			}
		}
		lead++
	}
}

// reveal reveals the given space.
func (m Matrix) reveal(space *field.Space) error {
	for _, row := range m {
		for _, cell := range row.Lhs() {
			if cell.space == space {
				cell.value = 0
				cell.state = field.Revealed
			}
		}
		if err := row.validate(); err != nil {
			return err
		}
	}
	return nil
}

// flag flags the given space.
func (m Matrix) flag(space *field.Space) error {
	for _, row := range m {
		for _, cell := range row.Lhs() {
			if cell.space == space {
				row.Rhs().value -= cell.value
				cell.value = 0
				cell.state = field.Flagged
			}
		}
		if err := row.validate(); err != nil {
			return err
		}
	}
	return nil
}

// removeUnconstrained removes rows with all zero left-hand-sides and columns
// with unconstrained spaces. This is used to clean up the output and is not
// algorithmically necessary.
func (m Matrix) removeUnconstrained() Matrix {
	if len(m) == 0 {
		return m
	}
	return m.removeRows(m.zeroRows()).removeCols(m.unconstrainedSpaces())
}

// copy returns a deep copy of the matrix.
func (m Matrix) copy() Matrix {
	var matrix Matrix
	for _, row := range m {
		var newRow Row
		for _, cell := range row {
			newRow = append(newRow, cell.copy())
		}
		matrix = append(matrix, newRow)
	}
	return matrix
}

// removeRow removes the given row from the matrix.
func (m Matrix) removeRow(row *field.Space) Matrix {
	return m.removeRows(map[*field.Space]bool{row: true})
}

// removeRows removes the given rows from the matrix.
func (m Matrix) removeRows(rows map[*field.Space]bool) Matrix {
	var matrix Matrix
	for _, row := range m {
		if !rows[row.Rhs().space] {
			matrix = append(matrix, row)
		}
	}
	return matrix
}

// removeCol removes the given spaces from the matrix.
func (m Matrix) removeCols(cols map[*field.Space]bool) Matrix {
	var matrix Matrix
	for _, row := range m {
		var newRow Row
		for _, cell := range row {
			if !cols[cell.space] {
				newRow = append(newRow, cell)
			}
		}
		matrix = append(matrix, newRow)
	}
	return matrix
}

File: src/matrix/matrix.rs
use super::cell::Cell;
use super::row::Row;
use crate::field::{constants::EPSILON, Field, SpaceIdx, State};
use std::collections::{HashMap, HashSet};

pub struct Matrix {
    pub rows: Vec<Row>,
}

pub struct Solution {
    pub flagged: Vec<SpaceIdx>,
    pub revealed: Vec<SpaceIdx>,
}

impl Matrix {
    pub fn new(field: &Field) -> Self {
        let revealed_edge_spaces: Vec<_> = field.revealed_edge_spaces().collect();
        let unknown_edge_spaces: Vec<_> = field.unknown_edge_spaces().collect();

        let mut rows = Vec::new();

        for &revealed_idx in &revealed_edge_spaces {
            let mut cells = Vec::new();
            let unknown_neighbors: HashSet<_> = field.spaces()[revealed_idx.get()]
                .unknown_neighbors()
                .iter()
                .copied()
                .collect();

            for &unknown_idx in &unknown_edge_spaces {
                let value = if unknown_neighbors.contains(&unknown_idx) {
                    1.0
                } else {
                    0.0
                };
                cells.push(Cell::new(value, unknown_idx, State::Unknown));
            }

            let space = &field.spaces()[revealed_idx.get()];
            let rhs = (space.mine_neighbor_count().unwrap_or(0) as i16
                - space.flagged_neighbor_count() as i16) as f64;
            cells.push(Cell::new(rhs, revealed_idx, State::Unknown));

            rows.push(Row::new(cells));
        }

        let mut matrix = Self { rows };
        matrix.reduce();
        matrix = matrix.remove_unconstrained();
        matrix
    }

    pub fn split_coupled(&self) -> Vec<Matrix> {
        let mut matrices = Vec::new();
        let mut src = self.copy();

        while !src.rows.is_empty() {
            let first_row = src.rows.remove(0);
            let mut dst = Matrix {
                rows: vec![first_row],
            };

            Self::coupled_rows(&mut src, &mut dst);
            dst = dst.remove_unconstrained();
            matrices.push(dst);
        }

        matrices
    }

    pub fn resolve(&self) -> Result<Matrix, String> {
        let mut matrix = self.copy();

        // Collect operations to perform
        let mut operations = Vec::new();

        for row in &matrix.rows {
            let lhs = row.constrained_lhs();
            let mut lhs_negative_sum = 0.0;
            let mut lhs_positive_sum = 0.0;

            for cell in &lhs {
                if cell.value < 0.0 {
                    lhs_negative_sum += cell.value;
                } else {
                    lhs_positive_sum += cell.value;
                }
            }

            let rhs = row.rhs().value;

            if (rhs - lhs_negative_sum).abs() < EPSILON {
                for cell in row.lhs() {
                    if cell.value < 0.0 {
                        operations.push((cell.space_idx, State::Flagged));
                    } else if cell.value != 0.0 {
                        operations.push((cell.space_idx, State::Revealed));
                    }
                }
            } else if (rhs - lhs_positive_sum).abs() < EPSILON {
                for cell in row.lhs() {
                    if cell.value < 0.0 {
                        operations.push((cell.space_idx, State::Revealed));
                    } else if cell.value != 0.0 {
                        operations.push((cell.space_idx, State::Flagged));
                    }
                }
            }
        }

        // Apply operations
        for (space_idx, state) in operations {
            match state {
                State::Flagged => matrix.flag(space_idx)?,
                State::Revealed => matrix.reveal(space_idx)?,
                State::Unknown => {}
            }
        }

        Ok(matrix)
    }

    pub fn solve(&self, visualize: bool) -> Vec<Solution> {
        let space_idx = self.most_constrained_space();

        if space_idx.is_none() {
            let mut solution = Solution {
                flagged: Vec::new(),
                revealed: Vec::new(),
            };

            if !self.rows.is_empty() {
                for cell in self.rows[0].lhs() {
                    match cell.state {
                        State::Flagged => solution.flagged.push(cell.space_idx),
                        State::Revealed => solution.revealed.push(cell.space_idx),
                        State::Unknown => {}
                    }
                }
            }

            if visualize {
                self.print();
            }

            return vec![solution];
        }

        let mut solutions = Vec::new();
        let space_idx = space_idx.unwrap();

        // Try revealing
        let mut reveal_matrix = self.copy();
        if reveal_matrix.reveal(space_idx).is_ok() {
            if let Ok(resolved) = reveal_matrix.resolve() {
                solutions.extend(resolved.solve(visualize));
            }
        }

        // Try flagging
        let mut flag_matrix = self.copy();
        if flag_matrix.flag(space_idx).is_ok() {
            if let Ok(resolved) = flag_matrix.resolve() {
                solutions.extend(resolved.solve(visualize));
            }
        }

        solutions
    }

    pub fn constrained_spaces(&self) -> HashSet<SpaceIdx> {
        let mut spaces = HashSet::new();
        for row in &self.rows {
            for cell in row.constrained_lhs() {
                spaces.insert(cell.space_idx);
            }
        }
        spaces
    }

    pub fn print(&self) {
        if self.rows.is_empty() {
            println!("(empty matrix)");
            return;
        }

        // Print header
        print!(" ");
        for cell in self.rows[0].lhs() {
            print!("{:4} ", cell.space_idx.get());
        }
        println!();

        for cell in self.rows[0].lhs() {
            let state_char = match cell.state {
                State::Unknown => "U",
                State::Flagged => "F",
                State::Revealed => "R",
            };
            print!("    {}", state_char);
        }
        println!();

        // Print rows
        for row in &self.rows {
            for (i, cell) in row.cells.iter().enumerate() {
                if i == row.cells.len() - 1 {
                    print!(
                        " | {:4} {}",
                        format!("{:.1}", cell.value),
                        cell.space_idx.get()
                    );
                } else {
                    print!(" {:4}", format!("{:.1}", cell.value));
                }
            }
            println!();
        }
        println!();
    }

    fn most_constrained_space(&self) -> Option<SpaceIdx> {
        let mut most_constrained_space = None;
        let mut most_constrained_count = 0;
        let mut constraint_counts: HashMap<SpaceIdx, usize> = HashMap::new();

        for row in &self.rows {
            for cell in row.constrained_lhs() {
                let idx = cell.space_idx;
                *constraint_counts.entry(idx).or_insert(0) += 1;
                if constraint_counts[&idx] > most_constrained_count {
                    most_constrained_space = Some(idx);
                    most_constrained_count = constraint_counts[&idx];
                }
            }
        }

        most_constrained_space
    }

    fn unconstrained_spaces(&self) -> HashSet<SpaceIdx> {
        if self.rows.is_empty() {
            return HashSet::new();
        }

        let mut spaces = HashSet::new();
        for cell in self.rows[0].lhs() {
            spaces.insert(cell.space_idx);
        }

        for row in &self.rows {
            for cell in row.constrained_lhs() {
                spaces.remove(&cell.space_idx);
            }
        }

        spaces
    }

    fn zero_rows(&self) -> HashSet<SpaceIdx> {
        let mut zero_rows = HashSet::new();
        for row in &self.rows {
            if row.constrained_lhs().is_empty() {
                zero_rows.insert(row.rhs().space_idx);
            }
        }
        zero_rows
    }

    fn coupled_rows(src: &mut Matrix, dst: &mut Matrix) {
        if src.rows.is_empty() {
            return;
        }

        let dst_constrained = dst.constrained_spaces();

        let mut to_move = Vec::new();
        for (i, row) in src.rows.iter().enumerate() {
            for cell in row.constrained_lhs() {
                if dst_constrained.contains(&cell.space_idx) {
                    to_move.push(i);
                    break;
                }
            }
        }

        if !to_move.is_empty() {
            for &i in to_move.iter().rev() {
                dst.rows.push(src.rows.remove(i));
            }
            Self::coupled_rows(src, dst);
        }
    }

    fn reduce(&mut self) {
        if self.rows.is_empty() {
            return;
        }

        let mut lead = 0;
        let row_count = self.rows.len();
        let col_count = self.rows[0].cells.len();

        for r in 0..row_count {
            if lead >= col_count {
                return;
            }

            let mut i = r;
            while self.rows[i].cells[lead].value == 0.0 {
                i += 1;
                if i == row_count {
                    i = r;
                    lead += 1;
                    if lead == col_count {
                        return;
                    }
                }
            }

            self.rows.swap(i, r);

            let div = self.rows[r].cells[lead].value;
            for cell in &mut self.rows[r].cells {
                cell.value /= div;
            }

            for j in 0..row_count {
                if j != r {
                    let sub = self.rows[j].cells[lead].value;
                    for k in 0..col_count {
                        let val = self.rows[r].cells[k].value * sub;
                        self.rows[j].cells[k].value -= val;
                    }
                }
            }

            lead += 1;
        }
    }

    fn reveal(&mut self, space_idx: SpaceIdx) -> Result<(), String> {
        for row in &mut self.rows {
            let len = row.cells.len();
            for cell in &mut row.cells[..len - 1] {
                if cell.space_idx == space_idx {
                    cell.value = 0.0;
                    cell.state = State::Revealed;
                }
            }
            row.validate()?;
        }
        Ok(())
    }

    fn flag(&mut self, space_idx: SpaceIdx) -> Result<(), String> {
        for row in &mut self.rows {
            let mut rhs_adjustment = 0.0;
            let len = row.cells.len();
            for cell in &mut row.cells[..len - 1] {
                if cell.space_idx == space_idx {
                    rhs_adjustment = cell.value;
                    cell.value = 0.0;
                    cell.state = State::Flagged;
                }
            }
            row.rhs_mut().value -= rhs_adjustment;
            row.validate()?;
        }
        Ok(())
    }

    fn remove_unconstrained(self) -> Matrix {
        if self.rows.is_empty() {
            return self;
        }
        let zero_rows = self.zero_rows();
        let unconstrained_spaces = self.unconstrained_spaces();
        self.remove_rows(&zero_rows)
            .remove_cols(&unconstrained_spaces)
    }

    fn copy(&self) -> Matrix {
        Matrix {
            rows: self
                .rows
                .iter()
                .map(|row| Row::new(row.cells.clone()))
                .collect(),
        }
    }

    fn remove_rows(self, rows: &HashSet<SpaceIdx>) -> Matrix {
        Matrix {
            rows: self
                .rows
                .into_iter()
                .filter(|row| !rows.contains(&row.rhs().space_idx))
                .collect(),
        }
    }

    fn remove_cols(self, cols: &HashSet<SpaceIdx>) -> Matrix {
        Matrix {
            rows: self
                .rows
                .into_iter()
                .map(|row| {
                    Row::new(
                        row.cells
                            .into_iter()
                            .filter(|cell| !cols.contains(&cell.space_idx))
                            .collect(),
                    )
                })
                .collect(),
        }
    }
}
